File: LeetCode/!0046. Permutations.cpp
/*
0046. Permutations

给一组互不相同的数，返回他们所有可能的排列顺序。

思路：回溯法。
将数组分为两个部分：已经排列的部分（设[0, first-1]）和还没排列的部分（设为[first, n-1]）。
设当前要排列的数为first位，待填写的数下标为i，那么就将first和i位的数相交换，这样就当作[0, first]完成了交换，剩下没交换的位置位[first+1, n-1]。
然后在此基础上进行递归。
递归后将first和i位的数位置还原，再准备将first位与i+1位的数进行交换。
*/

#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> res;

    void backtrack(vector<int>& nums, int start, int len) {
        if (start == len) {
            res.emplace_back(nums);
            return;
        }

        for (int i = start; i < len; ++i) {
            swap(nums[i], nums[start]);
            backtrack(nums, start + 1, len);
            swap(nums[i], nums[start]);
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        int len = nums.size();
        backtrack(nums, 0, len);
        return res;
    }
};

File: LeetCode/!0046. Permutations.py
'''
0046. Permutations

给定一个 没有重复 数字的序列，返回其所有可能的全排列。

思路：回溯法，参考：
https://leetcode-cn.com/problems/permutations/solution/hui-su-suan-fa-python-dai-ma-java-dai-ma-by-liweiw/

自己写的，有一个地方有问题：就是应该用一个数组used，记录数字是否用过。其他和参考答案思路基本一致。
'''

from typing import List


class Solution:
    def backtrack(self, nums, used, curPermutation, res):
        # 当前序列的长度等于所有候选数字的个数
        if (len(curPermutation) == len(nums)):
            res.append(curPermutation.copy())
            return
        # 当前序列长度不足，继续找组合
        for i in range(len(nums)):
            if not used[i]:
                used[i] = True
                curPermutation.append(nums[i])
                # 递归找排列
                self.backtrack(nums, used, curPermutation, res)
                # 回溯
                used[i] = False
                curPermutation.pop()

    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        curPermutation = []
        # used列表，用于记录nums中数字是否被使用了
        used = [False for _ in range(len(nums))]
        self.backtrack(nums, used, curPermutation, res)
        return res
